test/Transforms/loop-fusion.mlir

```diff
@@ -2388,6 +2389,29 @@ func @mul_add_0(%arg0: memref<3x4xf32>, %arg1: memref<4x3xf32>, %arg2: memref<3x
   // CHECK-NEXT:   }
   // CHECK-NEXT: }
   // CHECK-NEXT: return
+  return
+}
+
+// Verify that 'fuseProducerConsumerNodes' doesn't fuse a producer loop with
+// multiple outgoing edges. Sibling loop fusion is disabled to properly test
+// producer-consumer fusion in isolation.
+// CHECK-LABEL: func @should_not_fuse_multi_outgoing_edge_store_producer
+func @should_not_fuse_multi_outgoing_edge_store_producer() {
```

Comment:
Producer consumer fusion by itself is expected to fuse everything here. You shouldn't need sibling fusion to be turned on for fusion here. As such, the desired output even with -sibling-fusion=false would be the fused nest. While the fused nest could be seen as a sibling fusion followed by producer consumer fusion (which is probably what you had in mind?), it can also be viewed just as a sequence of two producer consumer fusions. 

Disabling sibling fusion would be expected to stop fusion in other cases like these.

```
for %i =  ...
   affine.load %A[%i]
   affine.store %B[%i]

for %j = ...
   affine.load %A[%i]
   affine.store %C[%i]
```

There is no producer consumer relationship or reuse, but "sibling fusion" (being for input/RAR reuse) will fuse these.

Reply:
Yes, that makes sense to me. A mature producer-consumer fusion algorithm should fuse loops below. However, what I'm trying to test here is that the current producer-consumer algorithm is not fusing a producer with more than one outgoing edge because that is not supported currently (note the *three* `NOSIBLING` checks). If I didn't disable sibling fusion, producer-consumer algorithm wouldn't fuse the loops but sibling fusion would and I couldn't be sure who did it.

Reply:
Can you try a unit test as follows. This test would not allow sibling fusion of the two consumer loop nests because of the dependence on '%b'. 

// producer loop
for ...
store %v, %a[%i]

// Consumer loop 1
for ...
load %a[%i]
store %v, %b[%i]

// Consumer loop 2
for ...
load %a[%i]
load %b[%i]

Reply:
Thanks! I’ve been playing with the suggested test case:
```
func @should_not_fuse_multi_outgoing_edge_store_producer2(%a : memref<1xf32>, %b : memref<1xf32>) {
    %cst = constant 0.000000e+00 : f32
    affine.for %arg0 = 0 to 1 {
      affine.store %cst, %a[%arg0] : memref<1xf32>
    }
 
    affine.for %arg0 = 0 to 1 {
      %0 = affine.load %a[%arg0] : memref<1xf32>
      affine.store %cst, %b[%arg0] : memref<1xf32>
    }

    affine.for %arg0 = 0 to 1 {
      %0 = affine.load %a[%arg0] : memref<1xf32>
      %1 = affine.load %b[%arg0] : memref<1xf32>
    }
    return
  } 
```
Producer-consumer fusion fuses the three loops. It fuses 2nd and 3rd first, then 1st becomes a producer with a single outgoing edge and it’s fused later so we cannot use this test.
However, I noticed that if I make ‘%a’ and external memref in my original test case, I can test exactly what I want and sibling fusion doesn’t kick in. I think we can proceed this way and remove `clSiblingFusion` for now, although I kind of like that flag. It's useful for debugging/independent testing.